server/balancer.go

```diff
@@ -82,98 +94,168 @@ func (s *storageBalancer) GetResourceKind() ResourceKind {
 	}
 
 	return newTransferPeer(region, oldPeer, newPeer)
 }
 
-// replicaChecker ensures region has enough replicas.
+// replicaChecker ensures region has the best replicas.
 type replicaChecker struct {
-	cluster  *clusterInfo
-	opt      *scheduleOption
-	selector Selector
+	opt     *scheduleOption
+	rep     *Replication
+	cluster *clusterInfo
+	filters []Filter
 }
 
-func newReplicaChecker(cluster *clusterInfo, opt *scheduleOption) *replicaChecker {
+func newReplicaChecker(opt *scheduleOption, cluster *clusterInfo) *replicaChecker {
 	var filters []Filter
 	filters = append(filters, newStateFilter(opt))
 	filters = append(filters, newSnapshotCountFilter(opt))
 
 	return &replicaChecker{
-		cluster:  cluster,
-		opt:      opt,
-		selector: newBalanceSelector(storageKind, filters),
+		opt:     opt,
+		rep:     opt.GetReplication(),
+		cluster: cluster,
+		filters: filters,
 	}
 }
 
 func (r *replicaChecker) Check(region *regionInfo) Operator {
-	// If we have bad peer, we remove it first.
-	for _, peer := range r.collectBadPeers(region) {
-		return newRemovePeer(region, peer)
+	if op := r.checkDownPeer(region); op != nil {
+		return op
+	}
+	if op := r.checkOfflinePeer(region); op != nil {
+		return op
 	}
 
-	stores := r.cluster.getRegionStores(region)
-
-	// Remove redundant replicas.
-	if len(stores) > r.opt.GetMaxReplicas() {
-		source := r.selector.SelectSource(stores)
-		return newRemovePeer(region, region.GetStorePeer(source.GetId()))
+	if len(region.GetPeers()) < r.rep.GetMaxReplicas() {
+		newPeer, _ := r.selectBestPeer(region)
+		if newPeer == nil {
+			return nil
+		}
+		return newAddPeer(region, newPeer)
 	}
 
-	// Ensure enough replicas.
-	if len(stores) < r.opt.GetMaxReplicas() {
-		return r.addPeer(region)
+	if len(region.GetPeers()) > r.rep.GetMaxReplicas() {
+		oldPeer, _ := r.selectWorstPeer(region)
+		if oldPeer == nil {
+			return nil
+		}
+		return newRemovePeer(region, oldPeer)
 	}
 
-	return nil
+	return r.checkBetterPeer(region)
 }
 
-func (r *replicaChecker) addPeer(region *regionInfo, filters ...Filter) Operator {
-	stores := r.cluster.getStores()
+// selectBestPeer returns the best peer in other stores.
+func (r *replicaChecker) selectBestPeer(region *regionInfo, filters ...Filter) (*metapb.Peer, int) {
+	filters = append(filters, r.filters...)
+	filters = append(filters, newExcludedFilter(nil, region.GetStoreIds()))
 
-	excluded := newExcludedFilter(nil, region.GetStoreIds())
-	target := r.selector.SelectTarget(stores, append(filters, excluded)...)
-	if target == nil {
-		return nil
+	var (
+		maxStore *storeInfo
+		maxScore int
+	)
+
+	// Find the store with maximum score.
+	regionStores := r.cluster.getRegionStores(region)
+	for _, store := range r.cluster.getStores() {
+		if filterTarget(store, filters) {
+			continue
+		}
+		score := r.rep.GetReplicaScore(regionStores, store)
+		if maxStore == nil || compareStoreScore(store, score, maxStore, maxScore) > 0 {
+			maxStore = store
+			maxScore = score
+		}
+	}
```

Comment:
The result store with the max `ReplicaScore` may have a very large resource use ratio, we may need to consider the 2 factors together.